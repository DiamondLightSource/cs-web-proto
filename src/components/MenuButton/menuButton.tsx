import React from "react";
import { writePv } from "../../hooks/useSubscription";

import { VType, VEnum } from "../../vtypes/vtypes";
import { vtypeToString, stringToVtype } from "../../vtypes/utils";
import {
  InferWidgetProps,
  PVComponent,
  PVWidget,
  PVWidgetPropType
} from "../Widget/widget";

export interface MenuButtonProps {
  connected: boolean;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  value?: VType;
  readonly: boolean;
  style?: {};
}

export const MenuButtonComponent = (props: MenuButtonProps): JSX.Element => {
  let { connected, value = null, style = { color: "#000000" } } = props;

  // Store whether component is disabled or not
  let disabled = false;

  const defaultText = "Waiting for value";

  let options = [defaultText];
  // Using value to dictate displayed value as described here: https://reactjs.org/docs/forms.html#the-select-tag
  // Show 0 by default where there is only one option
  let displayIndex = 0;

  let readOnlyStyle = {};
  if (props.readonly === true) {
    disabled = true;
    readOnlyStyle = { cursor: "not-allowed" };
  }

  if (!connected || value === null) {
    disabled = true;
  } else if (value instanceof VEnum) {
    options = value.getDisplay().getChoices();
    displayIndex = value.getIndex();
  } else {
    options = [vtypeToString(value)];
    disabled = true;
  }

  const mappedOptions = options.map(
    (text, index): JSX.Element => {
      return (
        <option key={index} value={index}>
          {text}
        </option>
      );
    }
  );

  return (
    <select
      value={displayIndex}
      disabled={disabled}
      style={{ width: "100%", ...readOnlyStyle, ...style }}
      onChange={props.onChange}
    >
      {mappedOptions}
    </select>
  );
};

// Menu button which also knows how to write to a PV
export const SmartMenuButton = (props: PVComponent): JSX.Element => {
  // Function to send the value on to the PV
  function onChange(event: React.ChangeEvent<HTMLSelectElement>): void {
    writePv(props.pvName, stringToVtype(event.currentTarget.value));
  }

  return (
    <MenuButtonComponent
      connected={props.connected}
      value={props.value}
      style={props.style}
      readonly={props.readonly}
      onChange={onChange}
    />
  );
};

export const MenuButton = (
  props: InferWidgetProps<typeof PVWidgetPropType>
): JSX.Element => <PVWidget baseWidget={SmartMenuButton} {...props} />;

MenuButton.propTypes = PVWidgetPropType;
