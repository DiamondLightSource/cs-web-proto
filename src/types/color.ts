import { createCanvas } from "canvas";
import log from "loglevel";
export class Color {
  private r: number;
  private g: number;
  private b: number;
  private a: number;

  public static WHITE = new Color(255, 255, 255);
  public static BLACK = new Color(0, 0, 0);
  public static RED = new Color(255, 0, 0);
  public static GREEN = new Color(0, 128, 0);
  public static BLUE = new Color(0, 0, 255);
  public static YELLOW = new Color(255, 255, 0);
  public static PURPLE = new Color(127, 0, 127);
  public static PINK = new Color(255, 192, 203);
  public static ORANGE = new Color(255, 165, 0);
  public static TRANSPARENT = new Color(0, 0, 0, 0);

  // Use the browser to compute colours from names.
  public static fromName(cssColorName: string): string {
    let ctx = document.createElement("canvas").getContext("2d");
    if (!ctx) {
      // Fall back to the canvas npm package.
      ctx = createCanvas(100, 100).getContext("2d");
    }
    if (ctx) {
      ctx.fillStyle = cssColorName;
      return ctx.fillStyle;
    } else {
      log.warn(
        `Failed to parse color name ${cssColorName} using canvas element.`
      );
      return "#000000";
    }
  }

  public static parse(cssColor: string): Color {
    let r = 0;
    let g = 0;
    let b = 0;
    if (cssColor.startsWith("#")) {
      let intRep = parseInt(cssColor.slice(1), 16);
      b = Math.floor(intRep % 256);
      intRep /= 256;
      g = Math.floor(intRep % 256);
      intRep /= 256;
      r = Math.floor(intRep);
    } else if (cssColor.startsWith("rgb")) {
      const parts = cssColor.match(/rgb\((.*), (.*), (.*)\)/);
      if (parts !== null) {
        r = parseInt(parts[1]);
        g = parseInt(parts[2]);
        b = parseInt(parts[3]);
      }
    } else if (cssColor === "transparent") {
      return Color.TRANSPARENT;
    } else {
      return this.parse(this.fromName(cssColor));
    }

    return new Color(r, g, b);
  }

  public constructor(r: number, g: number, b: number, a = 255) {
    if (r < 0 || r > 255) {
      throw new Error(`r value ${r} out of range`);
    } else if (g < 0 || g > 255) {
      throw new Error(`g value ${g} out of range`);
    } else if (b < 0 || b > 255) {
      throw new Error(`b value ${b} out of range`);
    }

    this.r = r;
    this.g = g;
    this.b = b;
    this.a = a;
  }

  public rgbaString(): string {
    return `rgba(${this.r}, ${this.g}, ${this.b}, ${this.a})`;
  }
}
