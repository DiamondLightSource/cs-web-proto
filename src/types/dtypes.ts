import { Alarm, ALARM_NONE } from "./alarm";
import { Display, DISPLAY_NONE } from "./display";

export class DTime {
  public datetime: Date;

  public constructor(datetime: Date) {
    this.datetime = datetime;
  }
}

export function dtimeNow(): DTime {
  return new DTime(new Date());
}

interface DTypeValue {
  stringValue?: string;
  doubleValue?: number;
  arrayValue?: number[];
}

export class DType {
  public value: DTypeValue;
  public time: DTime;
  public alarm?: Alarm;
  public display?: Display;

  public constructor(
    value: DTypeValue,
    alarm?: Alarm,
    time?: DTime,
    display?: Display
  ) {
    // TODO check for no value.
    this.value = value;
    this.alarm = alarm;
    this.time = time ?? dtimeNow();
    this.display = display;
  }

  public getStringValue(): string {
    if (this.value.stringValue) {
      return this.value.stringValue;
    } else if (this.value.doubleValue) {
      return this.value.doubleValue?.toString();
    } else if (this.value.arrayValue) {
      return this.value.arrayValue?.toString();
    } else {
      return "";
    }
  }

  public getDoubleValue(): number {
    // TODO what if not defined?
    if (this.value.doubleValue) {
      return this.value.doubleValue;
    } else {
      return NaN;
    }
  }

  public getAlarm(): Alarm {
    return this.alarm ?? ALARM_NONE;
  }
  public getTime(): DTime {
    return this.time;
  }
  public getDisplay(): Display | undefined {
    return this.display ?? DISPLAY_NONE;
  }

  public toString(): string {
    return this.getStringValue();
  }
}

export function dtypeToString(dtype?: DType): string {
  if (dtype) {
    return dtype.getStringValue();
  } else {
    return "";
  }
}

export function valueToDtype(
  value: any,
  alarm = ALARM_NONE,
  time = dtimeNow(),
  display = DISPLAY_NONE
): DType {
  const dvalue: DTypeValue = {};
  if (typeof value === "string") {
    dvalue.stringValue = value;
  } else if (typeof value === "number") {
    dvalue.doubleValue = value;
  } else {
    dvalue.doubleValue = 0;
  }
  return new DType(dvalue, alarm, time, display);
}
