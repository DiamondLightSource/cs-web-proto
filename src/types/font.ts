export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  private size: number;
  private style: FontStyle;
  private typeface: string;
  private name?: string;

  public constructor(
    size: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  public css(): object {
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    return {
      fontFamily: this.typeface,
      fontSize: `${this.size / 10}rem`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
  }
}
