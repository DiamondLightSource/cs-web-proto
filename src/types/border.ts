import { Color } from "./color";

export enum BorderStyle {
  None,
  Line,
  Dashed,
  Dotted,
  GroupBox
}

const CssBorders: { [key in BorderStyle]: string } = {
  [BorderStyle.None]: "none",
  [BorderStyle.Line]: "solid",
  [BorderStyle.Dashed]: "dashed",
  [BorderStyle.Dotted]: "dotted",
  [BorderStyle.GroupBox]: "ridge"
};

export class Border {
  private style: BorderStyle;
  private color: Color;
  // This will render a border width in pixels. I expect that we will revisit this
  // at some later point, possibly to allow or fix to rems.
  private width: number;

  public constructor(style: BorderStyle, color: Color, width: number) {
    this.style = style;
    this.color = color;
    this.width = width;
  }

  public css(): object {
    return {
      borderStyle: CssBorders[this.style],
      borderWidth: `${this.width}px`,
      borderColor: this.color.rgbaString()
    };
  }
}
