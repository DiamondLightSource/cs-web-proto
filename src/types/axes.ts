import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx ?? 0;
    this.autoScale = this.autoScale ?? true;
    this.autoScaleThreshold = this.autoScaleThreshold ?? 0.95;
    this.axisColor = this.axisColor ?? new Color("rgb(0, 0, 0");
    this.axisTitle = this.axisTitle ?? "";
    this.showGrid = this.showGrid ?? false;
    this.gridColor = this.gridColor ?? new Color("rgb(0, 0, 0");
    this.dashGridLine = this.dashGridLine ?? true;
    this.timeFormat = this.timeFormat ?? 0;
    this.scaleFormat = this.scaleFormat ?? "";
    this.scaleFont = this.scaleFont ?? new Font();
    this.titleFont = this.titleFont ?? new Font(10, FontStyle.Bold);
    this.visible = this.visible ?? true;
    this.logScale = this.logScale ?? false;
    this.leftBottomSide = this.leftBottomSide ?? true;
    this.maximum = this.maximum ?? 0;
    this.minimum = this.minimum ?? 100;
    // If axis has index 1 or above, assume y axis
    let axisBool = false;
    if (this.index > 0) {
      axisBool = true;
    }
    this.yAxis = this.yAxis ?? axisBool;
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
