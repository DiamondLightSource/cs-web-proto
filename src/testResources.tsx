import React from "react";
import { DType, DAlarm } from "./types/dtypes";
import { FileProvider, PageState, TabState } from "./fileContext";
import { render, RenderResult } from "@testing-library/react";
import { Provider } from "react-redux";
import { MacroContext } from "./types/macros";
import { csReducer, CsState } from "./redux/csState";
import { createStore } from "redux";
import { BrowserRouter as Router } from "react-router-dom";
import {
  OPEN_WEBPAGE,
  WidgetAction,
  WritePv,
  WRITE_PV
} from "./ui/widgets/widgetActions";

// Helper functions for dtypes.
export function ddouble(
  doubleValue: number,
  alarm: DAlarm = DAlarm.NONE
): DType {
  return new DType({ doubleValue: doubleValue }, alarm);
}

export function ddoubleArray(
  arrayValue: number[],
  alarm: DAlarm = DAlarm.NONE
): DType {
  return new DType({ arrayValue: Float64Array.from(arrayValue) }, alarm);
}

export function dstring(
  stringValue: string,
  alarm: DAlarm = DAlarm.NONE
): DType {
  return new DType({ stringValue: stringValue }, alarm);
}

// Test actions
export const WRITE_PV_ACTION: WritePv = {
  type: WRITE_PV,
  writePvInfo: {
    pvName: "PV",
    value: "value",
    description: "write value to PV"
  }
};

export const WRITE_PV_ACTION_NO_DESC: WritePv = {
  type: WRITE_PV,
  writePvInfo: {
    pvName: "PV",
    value: "value"
  }
};

export const OPEN_BBC_ACTION: WidgetAction = {
  type: OPEN_WEBPAGE,
  openWebpageInfo: { url: "https://bbc.co.uk", description: "BBC" }
};

export const ACTIONS_EX_AS_ONE = {
  actions: [WRITE_PV_ACTION, WRITE_PV_ACTION_NO_DESC],
  executeAsOne: true
};

export const ACTIONS_EX_FIRST = {
  actions: [WRITE_PV_ACTION, WRITE_PV_ACTION_NO_DESC],
  executeAsOne: false
};

// Helper function for rendering with a working fileContext.
export function contextRender(
  component: JSX.Element,
  initialPageState: PageState = {},
  initialTabState: TabState = {},
  initialCsState: CsState = {
    effectivePvNameMap: {},
    globalMacros: {},
    subscriptions: {},
    valueCache: {},
    deviceCache: {}
  }
): RenderResult {
  const ParentComponent = (props: { child: JSX.Element }): JSX.Element => {
    // Hard-code macros for now.
    // eslint-disable-next-line no-template-curly-in-string
    const contextMacros = { a: "A", b: "B", c: "C", e: "${a}" };
    const globalMacros = { c: "D", d: "E" };
    initialCsState.globalMacros = globalMacros;
    const macroContext = { macros: contextMacros, updateMacro: (): void => {} };
    const store = createStore<CsState, any, any, any>(
      csReducer,
      initialCsState
    );
    return (
      <Router>
        <Provider store={store}>
          <MacroContext.Provider value={macroContext}>
            <FileProvider
              initialPageState={initialPageState}
              initialTabState={initialTabState}
            >
              {props.child}
            </FileProvider>
          </MacroContext.Provider>
        </Provider>
      </Router>
    );
  };
  return render(<ParentComponent child={component}></ParentComponent>);
}
