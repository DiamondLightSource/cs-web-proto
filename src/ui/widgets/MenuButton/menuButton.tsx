import React, { CSSProperties, useContext } from "react";

import { commonCss, Widget } from "../widget";
import { PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  BoolProp,
  ColorPropOpt,
  InferWidgetProps,
  StringPropOpt
} from "../propTypes";
import { DType } from "../../../types/dtypes";
import {
  executeAction,
  getActionDescription,
  WidgetAction,
  WidgetActions,
  WritePv,
  WRITE_PV
} from "../widgetActions";
import { FileContext } from "../../../fileContext";
import { Border } from "../../../types/border";
import { Color } from "../../../types/color";

export interface MenuButtonProps {
  connected: boolean;
  onChange: (action: WidgetAction) => void;
  pvName?: string;
  value?: DType;
  readonly: boolean;
  actionsFromPv: boolean;
  label?: string;
  actions?: WidgetActions;
  foregroundColor?: Color;
  backgroundColor?: Color;
  border?: Border;
}

export const MenuButtonComponent = (props: MenuButtonProps): JSX.Element => {
  const {
    connected,
    value = null,
    readonly,
    actionsFromPv,
    pvName,
    label
  } = props;
  let actions: WidgetAction[] = props.actions?.actions || [];

  // Store whether component is disabled or not
  let disabled = readonly;

  let options: string[] = label ? [label] : [];
  const displayOffset = label ? 1 : 0;

  // Using value to dictate displayed value as described here: https://reactjs.org/docs/forms.html#the-select-tag
  // Show 0 by default where there is only one option
  let displayIndex = 0;

  if (actionsFromPv && pvName) {
    if (!connected || value === null) {
      disabled = true;
    } else if (value?.display?.choices) {
      options = options.concat(value?.display?.choices);
      actions = options.map((option, i) => {
        const writePv: WritePv = {
          type: WRITE_PV,
          writePvInfo: {
            pvName: pvName,
            value: i
          }
        };
        return writePv;
      });
      displayIndex = (value.getDoubleValue() ?? 0) + displayOffset;
    } else {
      disabled = true;
    }
  } else {
    options = options.concat(
      actions.map(action => {
        return getActionDescription(action);
      })
    );
  }

  const style: CSSProperties = {
    ...commonCss(props),
    width: "100%",
    height: "100%",
    textAlignLast: "center",
    cursor: disabled ? "not-allowed" : "default"
  };

  const mappedOptions = options.map((text, index): JSX.Element => {
    return (
      <option
        key={index}
        value={index}
        disabled={index === 0 && text === label}
      >
        {text}
      </option>
    );
  });

  /* Don't disable the element itself because that prevents
     any interaction even for ancestor elements, including middle-click copy. */
  function onMouseDown(event: React.MouseEvent<HTMLSelectElement>): void {
    if (disabled) {
      event.preventDefault();
    }
  }

  return (
    <select
      value={displayIndex}
      onMouseDown={onMouseDown}
      style={style}
      onChange={event => {
        props.onChange(
          actions[parseFloat(event.currentTarget.value) - displayOffset]
        );
      }}
    >
      {mappedOptions}
    </select>
  );
};

// Menu button which also knows how to write to a PV
export const SmartMenuButton = (props: {
  connected: boolean;
  pvName: string;
  value?: DType;
  readonly: boolean;
  actionsFromPv: boolean;
  actions?: WidgetActions;
  label?: string;
  foregroundColor?: Color;
  backgroundColor?: Color;
  border?: Border;
}): JSX.Element => {
  const files = useContext(FileContext);
  // Function to send the value on to the PV
  function onChange(action: WidgetAction): void {
    // The value from the select element is an integer as a string,
    // so we parse it into a float.
    executeAction(action, files);
  }

  return (
    <MenuButtonComponent
      pvName={props.pvName}
      connected={props.connected}
      value={props.value}
      readonly={props.readonly}
      actionsFromPv={props.actionsFromPv}
      actions={props.actions}
      onChange={onChange}
      label={props.label}
      foregroundColor={props.foregroundColor}
      backgroundColor={props.backgroundColor}
    />
  );
};

const MenuButtonWidgetProps = {
  ...PVWidgetPropType,
  actionsFromPv: BoolProp,
  label: StringPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt
};

export const MenuButton = (
  props: InferWidgetProps<typeof MenuButtonWidgetProps>
): JSX.Element => <Widget baseWidget={SmartMenuButton} {...props} />;

registerWidget(MenuButton, MenuButtonWidgetProps, "menubutton");
