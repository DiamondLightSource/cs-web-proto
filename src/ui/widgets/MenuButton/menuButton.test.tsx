import React from "react";
import { MenuButtonComponent } from "./menuButton";
import { create } from "react-test-renderer";
import { dtimeNow, DAlarm, DType, DDisplay } from "../../../types/dtypes";
import { ACTIONS_EX_FIRST, WRITE_PV_ACTION } from "../../../testResources";
import { fireEvent, render, waitFor } from "@testing-library/react";

const mock = jest.fn();
beforeEach((): void => {
  mock.mockReset();
});

function getMenubuttonComponent(
  value?: number,
  readonly?: boolean
): JSX.Element {
  const val = value ?? 0;
  const disabled = readonly ?? false;
  return (
    <MenuButtonComponent
      connected={true}
      value={
        new DType(
          { doubleValue: val },
          DAlarm.NONE,
          dtimeNow(),
          new DDisplay({
            choices: ["zero", "one", "two", "three", "four", "five"]
          })
        )
      }
      readonly={disabled}
      pvName="testpv"
      actionsFromPv={true}
      onChange={mock}
    />
  );
}

const menuButtonActions = (
  <MenuButtonComponent
    connected={false}
    readonly={false}
    actionsFromPv={false}
    actions={ACTIONS_EX_FIRST}
    label="menu button with label"
    onChange={mock}
  />
);
const menuButtonActionsDisabled = (
  <MenuButtonComponent
    connected={false}
    readonly={true}
    actionsFromPv={false}
    actions={ACTIONS_EX_FIRST}
    label="menu button with label"
    onChange={mock}
  />
);

describe("<MenuButton />", (): void => {
  test("it matches the snapshot", (): void => {
    const snapshot = create(getMenubuttonComponent());
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders all the choices", (): void => {
    const { getByRole } = render(getMenubuttonComponent());
    const select = getByRole("combobox");
    expect(select.childElementCount).toBe(6);
  });
  test("it renders actions", (): void => {
    const { getByRole } = render(menuButtonActions);
    const select = getByRole("combobox");
    expect(select.firstChild?.textContent).toEqual("menu button with label");
    // Two actions plus label.
    expect(select.childElementCount).toBe(3);
  });
  test("it renders the option with the correct index", (): void => {
    const { getByText } = render(getMenubuttonComponent(5));
    expect((getByText("zero") as HTMLOptionElement).selected).toBe(false);
    expect((getByText("five") as HTMLOptionElement).selected).toBe(true);
  });

  test("function called on click", async (): Promise<void> => {
    const { getByRole } = render(menuButtonActions);
    // Select the 1st option (0th option is the label)
    fireEvent.change(getByRole("combobox"), {
      target: { value: 1 }
    });
    expect(mock).toHaveBeenCalledWith(WRITE_PV_ACTION);
  });
  test("preventDefault called on mousedown when widget is disabled", async (): Promise<void> => {
    const { getByRole } = render(getMenubuttonComponent(0, true));
    const mockPreventDefault = jest.fn();
    const event = new MouseEvent("mousedown", { bubbles: true });
    event.preventDefault = mockPreventDefault;
    fireEvent(getByRole("combobox"), event);
    expect(mockPreventDefault).toHaveBeenCalled();
  });
});
