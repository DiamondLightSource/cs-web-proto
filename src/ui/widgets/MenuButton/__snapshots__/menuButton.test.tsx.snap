// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuButton /> it matches the snapshot 1`] = `
<select
  onChange={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "color": "#000000",
      "width": "100%",
    }
  }
  value={0}
>
  <option
    value={0}
  >
    zero
  </option>
  <option
    value={1}
  >
    one
  </option>
  <option
    value={2}
  >
    two
  </option>
  <option
    value={3}
  >
    three
  </option>
  <option
    value={4}
  >
    four
  </option>
  <option
    value={5}
  >
    five
  </option>
</select>
`;
